src/main/java/org/gridsuite/voltageinit/server/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,241 @@
+                        resultContext.getRunContext().getVariantId());
+                S result = run(network, resultContext.getRunContext(), resultContext.getResultUuid());
+
+                long nanoTime = System.nanoTime();
+                LOGGER.info("Just run in {}s", TimeUnit.NANOSECONDS.toSeconds(nanoTime - startTime.getAndSet(nanoTime)));
+
+                if (resultCanBeSaved(result)) {
+                    observer.observe("results.save", resultContext.getRunContext(), () -> saveResult(network, resultContext, result));
+
+                    long finalNanoTime = System.nanoTime();
+                    LOGGER.info("Stored in {}s", TimeUnit.NANOSECONDS.toSeconds(finalNanoTime - startTime.getAndSet(finalNanoTime)));
+
+                    sendResultMessage(resultContext, result);
+                    LOGGER.info("{} complete (resultUuid='{}')", getComputationType(), resultContext.getResultUuid());
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                if (!(e instanceof CancellationException)) {
+                    LOGGER.error(NotificationService.getFailedMessage(getComputationType()), e);
+                    notificationService.publishFail(
+                            resultContext.getResultUuid(), resultContext.getRunContext().getReceiver(),
+                            e.getMessage(), resultContext.getRunContext().getUserId(), getComputationType());
+                    resultService.delete(resultContext.getResultUuid());
+                    this.handleNonCancellationException(resultContext, e);
+                }
+            } finally {
+                futures.remove(resultContext.getResultUuid());
+                cancelComputationRequests.remove(resultContext.getResultUuid());
+            }
+        };
+    }
+
+    /**
+     * Handle exception in consumeRun that is not a CancellationException
+     * @param resultContext The context of the computation
+     * @param exception The exception to handle
+     */
+    protected void handleNonCancellationException(AbstractResultContext<R> resultContext, Exception exception) {
+    }
+
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> cancelAsync(CancelContext.fromMessage(message));
+    }
+
+    protected abstract void saveResult(Network network, AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void sendResultMessage(AbstractResultContext<R> resultContext, S result);
+
+    /**
+     * Do some extra task before running the computation, e.g. print log or init extra data for the run context
+     * @param ignoredRunContext This context may be used for further computation in overriding classes
+     */
+    protected void preRun(R ignoredRunContext) {
+        LOGGER.info("Run {} computation ...", getComputationType());
```

Comment:
no space at end in english
```suggestion
        LOGGER.info("Run {} computation...", getComputationType());
```